atlasdb-commons/src/main/java/com/palantir/common/compression/ClientCompressor.java

```
@@ -0,0 +1,71 @@
+
+import net.jpountz.lz4.LZ4BlockInputStream;
+
+public enum ClientCompressor {
+    GZIP(GzipCompressingInputStream.class, GZIPInputStream.class, GzipCompressingInputStream.GZIP_HEADER),
+    LZ4(LZ4CompressingInputStream.class, LZ4BlockInputStream.class,
+            new byte[] {'L', 'Z', '4', 'B', 'l', 'o', 'c', 'k'}),
+    NONE(null, null, new byte[] {});
+
+    public final Class<InputStream> compressorClass;
+    public final Class<InputStream> decompressorClass;
+    public final byte[] magic;
+
+    ClientCompressor(Class compressorClass, Class decompressorClass, byte[] magic) {
+        this.compressorClass = compressorClass;
+        this.decompressorClass = decompressorClass;
+        this.magic = magic;
+    }
+
+    private static boolean matchMagic(ClientCompressor compressorType, byte[] buffer, int bufferLen) {
+        byte[] signature = compressorType.magic;
+        int i = 0;
+        while (i < signature.length && i < bufferLen && signature[i] == buffer[i++]);
+        return i >= signature.length && signature.length > 0;
+    }
+
+    static InputStream getDecompressorStream(InputStream stream) throws IOException {
+        BufferedInputStream buff = new BufferedInputStream(stream);
+        int maxLen = Arrays.stream(ClientCompressor.values()).max(
+                Comparator.comparingInt(x -> x.magic.length)).get().magic.length;
+        buff.mark(maxLen);
+        byte[] headerBuffer = new byte[maxLen];
+        int len = buff.read(headerBuffer);
+        buff.reset();
+        Optional<ClientCompressor> compressor = Arrays.stream(ClientCompressor.values()).filter(t -> matchMagic(t, headerBuffer, len)).findFirst();
+        return compressor.map(t -> {
+            try {
+                return t.decompressorClass.getConstructor(InputStream.class).newInstance(buff);
+            } catch (Exception exc) {
+                throw new RuntimeException(exc);
+            }
+        }).orElseGet(() -> buff);
```

this isn't right - if it doesn't match you shouldn't just return the compressed information - you should throw an exception.

---

by returning the buff, It handles the case where the given stream is not a compression stream at all. So it is safe to pass a non compressed stream and read from it as it will just wrap it with buffered stream. We are not exposing the decompressor class to the user anyways.